### ISSUE TITLE
Some problems with ISO Loader images list and some requests

### ISSUE THREAD

**Issue by vsub** [PCSX2/pcsx2#877]
1.It keeps literally everything you try to run...even pictures,text,exe files and on.
Here is an example.
Start pcsx2 and run the bios(or a game) 
Drag drop anything over the main window
Pcsx2 will try to run it and if unsuccessful,an error message will appear and then pcsx2 will freeze...and add that item to the iso list(I tried with the pcsx2 exe and after I kill the process and start pcsx2 again,the pcsx2.exe was on the list)

You can also add an item if you use Open With or just drop something over the pcsx2.exe(which is the same thing as wanting to run a game image)

An item to the list should only be added on successful boot...you can call a successful boot if pcsx2 was able to get the CRC.

2.It display items even if they don't exist...can it be done to check the existence of the file before adding it to the menu?
For example if you run a game and then move\delete it,pcsx2 should not display it on the list.It can stay in the ini if it's useful for something but the list should contains only items that exist.

Selecting a game from a list full of items pointing to non existing files is not fast\easy 

3.The paths should be full not relative(add an ini option for it).
Relative and full path have it's uses but I have more problems with relative than full paths....it would be nice to have an ini option to change between using relative or full paths.

**Comment by ssakash** [comment]
>For example if you run a game and then move\delete it,pcsx2 should not display it on the list.

It'd be greyed out of the list which would probably mean it's unusable , you don't prefer that method ?

**Comment by vsub** [comment]
No,it just makes the list bigger and you don't have any use of that item until you correct the paths.

For example you run some image or elf file and then detele it...that item will forever stay there making the list bigger until you remove it manually

**Comment by avih** [comment]
- It's not easy to define "unsuccessful".

You ask pcsx2 to open something, and it tries to open it as an DVD image file. It might be a valid PS2 DVD image file which PCSX2 is just unable to run due to compatibility issue (e.g. maybe it won't even reach the menu, or reach the menu but fail to start the game, etc), or it might be a valid DVD image but not of a PS2 game, or it might be a corrupted DVD image file, or it might be corrupted in a way which the boot starts fine but then fails to read from the image at some stage, etc.

Which of those would you not like to stay at the Iso selector list?


- Display items which don't exist.

True, but they should be grayed out. This is by design, especially for portable PCSX2, such that if you put PCSX2 on a USB drive and take it to another system (where some ISOs are maybe not found) and then take it back to a system where they are found, then it will display them as disabled on the other system but display them as enabled again on the original system where they are found.

If you have better suggestions on how to handle it, shoot.

- Relative vs full paths.

It's really simple and not specific to the ISO images paths. If PCSX2 is installed using the installer, it uses absolute paths. If PCSX2 is portable, it uses relative paths (to pcsx2.exe) as much as possible. It's also somewhat smart about the paths in portable mode:
- If it's within the same dir as pcsx2.exe, then it's saved as relative to that dir.
- if it's outside the pcsx2.exe dir but at the same drive [letter], then it saves it relative to the drive root, e.g. `/my-isos/whatever.iso` (so you could put it on a portable drive but outside the pcsx2 dir and it will still find the ISOs).
- If it's at a different drive that where pcsx2.exe is, then it saves the full path.

We will not make it a configuration whether or not to use relative paths because pcsx2 configuration is already very very complex. Portable is relative, installed is absolute, that's it.

However, if you have concrete examples where the relative paths are an issue, please give them and we'll examine if there's a bug or whether the system could be improved.

**Comment by vsub** [comment]
> You ask pcsx2 to open something, and it tries to open it as an DVD image file. It might be a valid PS2 DVD image file which PCSX2 is just unable to run due to compatibility issue (e.g. maybe it won't even reach the menu, or reach the menu but fail to start the game, etc), or it might be a valid DVD image but not of a PS2 game, or it might be a corrupted DVD image file, or it might be corrupted in a way which the boot starts fine but then fails to read from the image at some stage, etc.

Umm which of those are actually useful to stay in the list if they all don't run.
As I suggested,execute the function that writes the image path to the ini only after pcsx2 get the crc.I can't think of a reason why such image should not be on the list.

> If you have better suggestions on how to handle it, shoot.

As I suggested(I don't know how pcsx2 build it list)but can it check if the file exist before adding it to the list.

For example:
Get the path from Filename00=,if it exist create a menu item for it,and if it doesn't skip to the next.
That way the list will contains only working items
I'm not talking about deleting items from the ini

> If PCSX2 is installed using the installer, it uses absolute paths. If PCSX2 is portable...

It's all good but it's not useful if you use the portable version and try to test some beta of pcsx2(on another drive)while using the same settings.
I like to have everything pcsx2 related in the pcsx2 folder and that's why I'm using the portable version

**Comment by avih** [comment]
> execute the function that writes the image path to the ini only after pcsx2 get the crc

These are implementation details which might not be relevant to the eventually chosen implementation, but your suggestion here is to only keep items at the list if booting them was able to get a CRC (probably more specifically, if it found an ELF to run). Agreed that it sounds useful. Not trivial to implement but useful.


> Get the path from Filename00=,if it exist create a menu item for it,and if it doesn't skip to the next.

It's a personal preference IMO (I like the grayed out items), and this doesn't sound like a bad suggestion, but might also be not trivial to implement since IIRC the number at the ini file is directly associated with the CDVD menu id number, and having the ini contain some items which are completely not displayed would make it harder to handle.

Also, since the list is limited in length (10 by default, can be modified at the INI), it could create weird looking scenarios where the list holds less items than it should (needs testing though).


> It's all good but it's not useful if you use the portable version and try to test some beta of pcsx2

Give me a concrete example please and I'll respond to it.

Also, portable apps are portable - you should be able to move them around and keep using them. I don't think there's a "standard" of how a portable application should behave when it needs to remember file locations outside of its portable folder, but I think PCSX2 handles this case quite well.

I think it's highly unlikely that portable installs of PCSX2 will always use absolute paths for every config path, or even just for ISO paths, since this would definitely break any possibility of moving the folder around and keeping it working.

Also do keep in mind that there are limits to how much configurability you can add to a program, and PCSX2 specifically already has a lot of configurations. You can't make the use absolute/relative path a configuration value because then pcsx2 wouldn't know where should it read its ini files from, for example. Or you could say "so only make it a configuration for ISO paths", but IMO it's just unlikely to happen.

**Comment by vsub** [comment]
> Also, since the list is limited in length (10 by default, can be modified at the INI), it could create weird looking scenarios where the list holds less items than it should (needs testing though).

Then can you at least make it to accept 0...I always wanted to always be asked what to load,not choosing it from the menu.
This is something personal but it will solve all of my problems with that menu and this issue can be closed

**Comment by avih** [comment]
> Then can you at least make it to accept 0...I always wanted to always be asked what to load,not choosing it from the menu.

Not sure I understand this. What would `0` mean (how would it behave)? and what do you mean by "I want to be asked"?

**Comment by vsub** [comment]
If the value is above 0,the iso list will keep that much items but if the list is empty,pcsx2 will always ask me what to load.
Currently pcsx2 doesn't accept 0,and even crash if you set 0 and try to run a game

**Comment by avih** [comment]
Again, what do you mean by "pcsx2 will ask me what to load"? Do you expect a Siri like voice to ask you?

**Comment by vsub** [comment]
Clear all of your recent games from the ini and try fast/full boot without choosing a game from the ISO Selector=>Browse

If no games exist in the list,pcsx2 will ask you what to load and then automatically run it.

**Comment by avih** [comment]
I don't get it, sorry. Maybe someone else would be able to help me understand what you mean.

**Comment by vsub** [comment]
1.Open the pcsx2_ui.ini(I think that was the name)
2.Delete everything from Recent ISO(all those Filename0?=)and save
3.Start pcsx2=>choose CDVD=>ISO Loader
4.System=>Boot CDVD (fast or full).

Doing that(skipping the CDVD=>ISO Selector=>Browse=>Choose Image),pcsx2 will open the Browse dialog and when you choose the image and hit ok,pcsx2 will directly run it.

This happens only if there is nothing in the ISo Selector list

**Comment by avih** [comment]
Let's ignore the ini file for now and what you do with it, since it's implementation details, and we don't expect users to edit the ini, right?

PCSX2 already opens the "Browse" dialog if you boot and there are no recent ISOs.

So you want to add a single new feature: a button or menu item which will make PCSX2 forget all the ISOs it knows about. Is that it?

**Comment by vsub** [comment]
> Let's ignore the ini file for now and what you do with it, since it's implementation details, and we don't expect users to edit the ini, right?

The point was to use existing feature rather than creating new one(you choose how much images to display and if it's set to 0,that will make pcsx2 to not create a list and this will result in "always asking what to load")

> So you want to add a single new feature: a button or menu item which will make PCSX2 forget all the ISOs it knows about, and after you click this button, and do boot, it will open the ISO selector dialog.

Well maybe add a menu with a check that will be called something like " don't remember recently played" or something like that and if that option is enabled,pcsx2 will not write to the ini and create a list and this will result in "always asking what to load" because when the is empty,pcsx2 ask what to load...like you said

> PCSX2 already opens the ISO selector dialog if you boot and there are no recent ISOs.

And that dialog has an extra feature when you run it like this...it directly run the game

I always liked how Linuzappz ISO worked(when you don't set an iso)and I want that feature in the ISo loader...to always ask me what to load and when I choose it,to directly run it.

Not like now first choose it from the list and then run it...I have big collection of games and having to first choose a game(not the game played last time)and then run it from another menu is kinda annoying.

Running a game when the list is empty is much faster.

**Comment by avih** [comment]
In general, when discussing bugs and improvements it's important to make distinction between few mostly unrelated areas:
- **The problem**. This is the most important thing to understand - and understand it before we get to the others. This should be the answer to the question of **What bothers you?** or **Why do you want this to change?**, i.e. what bothers someone with the current behavior? in what cases it doesn't do something good enough? or behaves unexpectedly, etc.
- **Suggested solutions**. The possible answers to the question of **What could/should it to do instead in terms of behavior and maybe also UI?** -  For the same problem there could be several solutions, where each of them tries to offer a different remedy to "the problem". Sometimes, however, especially for proper bugs, this would be an obvious "do X instead of Y", but that's not always the case.
- **Implementation details** - this should be left for the developer to decide and in general whoever is reporting the problem shouldn't be concerned with that (though obviously suggestions are welcome). It might include what values to keep at the ini file, or what parts of the code should run and when, etc.

However, when someone is requesting e.g. "keep 0 at the ini file" - it doesn't help the developer understand what's the problem it tries to solve, or what's the suggested solution someone has in mind. There could be better solutions, and there could be many different implementation details. So this makes the discussion and solving the problem much much harder.

Most of your posts mix different suggested solutions and different implementation details, and sometimes it's hard to understand which is which.

So far, from all the text in this issue, I recognize a single problem with two sub-problems:
- The problem is that the CDVD menu keeps items which you don't want to see. Specifically, you describe two categories of such items:
   - DVD Images which are no longer available on the system.
   - Items which you accidentally tried to open but which PCSX2 can't open.

Before we continue, let me know if I got "the problem" 100% right so far.

**Comment by refractionpcsx2** [comment]
Closing this as we're moving to QT so the "recent" list will no longer exist, you will have a game list instead.
